src/Avalonia.Base/Utilities/AvaloniaPropertyValueStore.cs

```diff
@@ -8,166 +7,197 @@ namespace Avalonia.Utilities
     /// Stores values with <see cref="AvaloniaProperty"/> as key.
     /// </summary>
     /// <typeparam name="TValue">Stored value type.</typeparam>
-    internal sealed class AvaloniaPropertyValueStore<TValue>
+    internal struct AvaloniaPropertyValueStore<TValue>
     {
-        // The last item in the list is always int.MaxValue.
-        private static readonly Entry[] s_emptyEntries = { new Entry { PropertyId = int.MaxValue, Value = default! } };
-        
-        private Entry[] _entries;
+        private Entry[]? _entries;
+        private int _entryCount;
 
         public AvaloniaPropertyValueStore()
         {
-            _entries = s_emptyEntries;
+            _entries = null;
+            _entryCount = 0;
+            IsInitializing = false;
+            InitialSize = 4;
         }
 
-        public int Count => _entries.Length - 1;
-        public TValue this[int index] => _entries[index].Value;
+        public int Count => _entryCount;
+        
+        public bool IsInitializing { get; set; }
+
+        public int InitialSize { get; set; }
+
+        public TValue this[int index] => _entries![index].Value;
 
-        private (int, bool) TryFindEntry(int propertyId)
+        private EntryIndex LookupEntry(int propertyId)
```

Comment:
I've noticed from benchmarking that returning `struct`s is slightly slower than using the try pattern (about 10% slower on x64). Not something you'd usually worry about, but because this PR is about performance and the try pattern would fit here, it might make sense to use it.